Behavioral tests for Tax commands

zuk3975 opened the pull request:
<!--
Thank you for contributing to the PrestaShop project! 

Please take the time to edit the "Answers" rows with the necessary information.

Check out our contribution guidelines to find out how to complete it:
https://devdocs.prestashop.com/1.7/contribute/contribution-guidelines/#pull-requests
 -->

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Adds behavioral tests for tax create/edit/delete actions. Also adds missing typecasting in GetTaxForEditingHandler
| Type?         | refacto
| Category?     | TE
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | #14480
| How to test?  | Tests can be executed by command: `./vendor/bin/behat -c tests/Integration/Behaviour/behat.yml -s tax`

<!-- Click the form's "Preview" button to make sure the table is functional in GitHub. Thank you! -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/14132)
<!-- Reviewable:end -->

zuk3975 committed:
Adds behat for tax creation

zuk3975 committed:
Adds tax edit/delete scenarios

zuk3975 committed:
Removes unused property

zuk3975 committed:
Some code optimization and typo correction

zuk3975 committed:
Added invalid fields cases. Some optimization

zuk3975 committed:
Adds tax behat command reminder and minor wording fix in error msg

PierreRambaud commented on tests/Integration/Behaviour/Features/Context/Domain/TaxFeatureContext.php:
```diff
@@ -0,0 +1,255 @@
+
+    public function __construct()
+    {
+        $this->defaultLangId = Configuration::get('PS_LANG_DEFAULT');
```
Can we use Adapter classes here?
It will be easier to update when we'll remove legacy classes

zuk3975 committed:
Switches using Adapter/Configuration instead of direct legacy

zuk3975 committed:
Fix typecasting in GetTaxForEditingHandler and add bulk actions behats

zuk3975 committed:
Removes unused class import

zuk3975 committed:
Optimizes behats, removes cases that should be covered with unit testing

matks commented on tests/Integration/Behaviour/Features/Context/Domain/TaxFeatureContext.php:
```diff
@@ -0,0 +1,281 @@
+        $expectedStatus = 'enable' === $action ? true : false;
+
+        $this->getCommandBus()->handle(new BulkToggleTaxStatusCommand(
+            array_map('intval', explode(',', $ids)),
```
Tricky but funny 👍

matks commented on tests/Integration/Behaviour/Features/Scenario/Tax/tax_management.feature:
```diff
@@ -0,0 +1,56 @@
+    And tax "tax-1" rate should be 0.150
+    And tax "tax-1" should be disabled
+
+  Scenario: Editing tax when providing only changeable property should be allowed
```
```suggestion
  Scenario: It is possible to modify only the name of a tax, nothing else
```
Let's write very story-like scenarios 😄

matks commented on tests/Integration/Behaviour/Features/Scenario/Tax/tax_management.feature:
```diff
@@ -0,0 +1,56 @@
+    When I edit tax "tax-1" with following properties:
+      | name         | my custom tax 300 |
+      | rate         | 0.15              |
+      | is_enabled   | 0                 |
```
If you use `PrimitiveUtils::castStringBooleanIntoBoolean()` on this table row, you can write

```suggestion
      | is_enabled   | 0                 |
```
which is easier to read

matks commented on tests/Integration/Behaviour/Features/Scenario/Tax/tax_management.feature:
```diff
@@ -0,0 +1,56 @@
+    Then tax "tax-1" should be enabled
+
+  Scenario: Disabling taxes status in bulk action
+    Given taxes with ids: "1,2,3" exists
```
Mmmm 😐 it's sad we have a lot of good scenarios talking about tax references (like `tax-1` although it could be interesting to use names even closer to reality like "TVA", "State Tax", "IRS tax") and suddenly we have scenarios talking about IDs (which are technical-oriented)

Cant we create 3 taxes before using the bulk ?

```
When I add new tax "tax-1" with following properties:
      | name         | my custom tax 501 |
      | rate         | 0.5               |
      | is_enabled   | 1                 |
And When I add new tax "tax-2" with following properties:
      | name         | my custom tax 502 |
      | rate         | 0.6               |
      | is_enabled   | 1                 |
And I add new tax "tax-3" with following properties:
      | name         | my custom tax 503 |
      | rate         | 0.7               |
      | is_enabled   | 1                 |
And I disable the following taxes: "tax-1", "tax-2", "tax-3"
```

And we retrieve the IDs required for the Command from the tax references

matks commented on tests/Integration/Behaviour/Features/Scenario/Tax/tax_management.feature:
```diff
@@ -0,0 +1,56 @@
+    When I enable taxes with ids: "1,2,3"
+    Then taxes with ids: "1,2,3" should be enabled
+
+  Scenario: Deleting tax
```
You can mix some actions together to avoid having too small scenarios 😉 it's authorized as it's integration tests. It's actually interesting to see whether a test works well when paired with another. So we can for example merge the "Toggling tax status" with the "Deleting tax" (run one after the other) or "Disabling taxes status in bulk action" with "Deleting taxes in bulk action"

We can also aim for more user-oriented scenario titles:
`Deleting taxes in bulk action` => `Deleting multiple taxes from a BO listing`

zuk3975 committed:
Wip refactoring too technical cases

zuk3975 committed:
Fix wrong array and optimize some conditions

zuk3975 committed:
Refactoring deleteTax behats

zuk3975 committed:
Adds some mixed-cases behat tests

sarjon commented:
___
*[tests/Integration/Behaviour/Features/Context/Domain/TaxFeatureContext.php, line 96 at r3](https://reviewable.io/reviews/prestashop/prestashop/14132#-LkTWLKTAUbJrjpFmnrL:-LkTW7No76bCxX6lVC64:bon74u4) ([raw file](https://github.com/prestashop/prestashop/blob/ff8e6924d76f8769966448f7d0a92cb108a7213e/tests/Integration/Behaviour/Features/Context/Domain/TaxFeatureContext.php#L96)):*
> ```PHP
> 
>     /**
>      * @When I toggle tax :taxReference status
> ```

When toggling you won't know the end result. How about `I enable/disable tax ...`?


<!-- Sent from Reviewable.io -->

sarjon commented:
___
*[tests/Integration/Behaviour/Features/Context/Domain/TaxFeatureContext.php, line 150 at r3](https://reviewable.io/reviews/prestashop/prestashop/14132#-LkTX6eR4hCX2VofW6k8:-LkTX6eR4hCX2VofW6k9:b-80ovsg) ([raw file](https://github.com/prestashop/prestashop/blob/ff8e6924d76f8769966448f7d0a92cb108a7213e/tests/Integration/Behaviour/Features/Context/Domain/TaxFeatureContext.php#L150)):*
> ```PHP
>     {
>         foreach (PrimitiveUtils::castStringArrayIntoArray($taxReferences) as $taxReference) {
>             $this->deleteTax($taxReference);
> ```

Shouldn't you use bulk delete command instead?


<!-- Sent from Reviewable.io -->

sarjon reviewed (commented):
> Reviewed 1 of 3 files at r1, 1 of 3 files at r2, 2 of 2 files at r3.
> *[Reviewable](https://reviewable.io/reviews/prestashop/prestashop/14132)* status: all files reviewed, 16 unresolved discussions (waiting on @zuk3975)

___
*[tests/Integration/Behaviour/Features/Context/Domain/TaxFeatureContext.php, line 160 at r3](https://reviewable.io/reviews/prestashop/prestashop/14132#-LkTXUE3-NvIHtqsrbjA:-LkTXUE3-NvIHtqsrbjB:b-iib9s1) ([raw file](https://github.com/prestashop/prestashop/blob/ff8e6924d76f8769966448f7d0a92cb108a7213e/tests/Integration/Behaviour/Features/Context/Domain/TaxFeatureContext.php#L160)):*
> ```PHP
>     {
>         foreach (PrimitiveUtils::castStringArrayIntoArray($taxReferences) as $taxReference) {
>             $this->assertTaxisDeleted($taxReference);
> ```

`TaxIsDeleted` 
___
*[tests/Integration/Behaviour/Features/Context/Domain/TaxFeatureContext.php, line 167 at r3](https://reviewable.io/reviews/prestashop/prestashop/14132#-LkTY0gN19w7YkNF1zAK:-LkTY0gN19w7YkNF1zAL:b-r52r64) ([raw file](https://github.com/prestashop/prestashop/blob/ff8e6924d76f8769966448f7d0a92cb108a7213e/tests/Integration/Behaviour/Features/Context/Domain/TaxFeatureContext.php#L167)):*
> ```PHP
>      * @Then tax :taxReference should be deleted
>      */
>     public function assertTaxisDeleted($taxReference)
> ```

Same here, "is" camel case?
___
*[tests/Integration/Behaviour/Features/Context/Domain/TaxFeatureContext.php, line 224 at r3](https://reviewable.io/reviews/prestashop/prestashop/14132#-LkTYiT64ksrm4O5XvXk:-LkTYiT64ksrm4O5XvXl:b-eqt5ej) ([raw file](https://github.com/prestashop/prestashop/blob/ff8e6924d76f8769966448f7d0a92cb108a7213e/tests/Integration/Behaviour/Features/Context/Domain/TaxFeatureContext.php#L224)):*
> ```PHP
>     public function assertTaxesExistsByIds($ids)
>     {
>         foreach (explode(',', $ids) as $id) {
> ```

How about using `PrimitiveUtils`?
___
*[tests/Integration/Behaviour/Features/Scenario/Tax/tax_management.feature, line 4 at r3](https://reviewable.io/reviews/prestashop/prestashop/14132#-LkTZAd12orLJOH_pC3y:-LkTZAd12orLJOH_pC3z:bsnnquw) ([raw file](https://github.com/prestashop/prestashop/blob/ff8e6924d76f8769966448f7d0a92cb108a7213e/tests/Integration/Behaviour/Features/Scenario/Tax/tax_management.feature#L4)):*
> ```Gherkin
> #./vendor/bin/behat -c tests/Integration/Behaviour/behat.yml -s tax
> Feature: Manage tax
>   As a employee
> ```

As an employee
___
*[tests/Integration/Behaviour/Features/Scenario/Tax/tax_management.feature, line 5 at r3](https://reviewable.io/reviews/prestashop/prestashop/14132#-LkTZDhgAqLvpDX-HVa3:-LkTZDhgAqLvpDX-HVa4:b-mebd0g) ([raw file](https://github.com/prestashop/prestashop/blob/ff8e6924d76f8769966448f7d0a92cb108a7213e/tests/Integration/Behaviour/Features/Scenario/Tax/tax_management.feature#L5)):*
> ```Gherkin
> Feature: Manage tax
>   As a employee
>   I must be able to correctly add, edit and delete tax
> ```

Not sure "correctly" says anything.
___
*[tests/Integration/Behaviour/Features/Scenario/Tax/tax_management.feature, line 33 at r3](https://reviewable.io/reviews/prestashop/prestashop/14132#-LkTZnwq0ZRGPu4Sb6ta:-LkTZnwq0ZRGPu4Sb6tb:b-uvi0en) ([raw file](https://github.com/prestashop/prestashop/blob/ff8e6924d76f8769966448f7d0a92cb108a7213e/tests/Integration/Behaviour/Features/Scenario/Tax/tax_management.feature#L33)):*
> ```Gherkin
> 
>   Scenario: Toggling tax status
>     Given tax "sales-tax" should be disabled
> ```

"should be" gives impression that it can also be enabled. How about `Given tax "sales-tax" is disabled`?
___
*[tests/Integration/Behaviour/Features/Scenario/Tax/tax_management.feature, line 43 at r3](https://reviewable.io/reviews/prestashop/prestashop/14132#-LkT_TtP5tAAJFpWzfmJ:-LkT_TtP5tAAJFpWzfmK:brb06it) ([raw file](https://github.com/prestashop/prestashop/blob/ff8e6924d76f8769966448f7d0a92cb108a7213e/tests/Integration/Behaviour/Features/Scenario/Tax/tax_management.feature#L43)):*
> ```Gherkin
>     Then tax "sales-tax" should be deleted
> 
>   Scenario: Disabling multiple taxes status in bulk action
> ```

You can remove word "status" imo.
___
*[tests/Integration/Behaviour/Features/Scenario/Tax/tax_management.feature, line 58 at r3](https://reviewable.io/reviews/prestashop/prestashop/14132#-LkT_d0GDZDkFHp8fJ85:-LkT_d0GDZDkFHp8fJ86:b-dsea1t) ([raw file](https://github.com/prestashop/prestashop/blob/ff8e6924d76f8769966448f7d0a92cb108a7213e/tests/Integration/Behaviour/Features/Scenario/Tax/tax_management.feature#L58)):*
> ```Gherkin
>     When I disable taxes: "beard-tax, state-tax, pvm-tax"
>     Then taxes: "beard-tax, state-tax" should be disabled
>     And tax "pvm-tax" should be disabled
> ```

What's the point of checking this tax separately?


<!-- Sent from Reviewable.io -->

zuk3975 committed:
Removes status "toggling" - "enable/disable" instead. Removes unused method and redundant code

zuk3975 commented:
@matks addressed your comments. Travis is mad, he is dying to see `#blockcart-modal` (Error: element (#blockcart-modal) still not visible after 2500ms), which i guess is not up to me.

matks commented:
> @matks addressed your comments. Travis is mad, he is dying to see `#blockcart-modal` (Error: element (#blockcart-modal) still not visible after 2500ms), which i guess is not up to me.

Yes, these tests are flawed 😞 
I'm working hard to convince my teammates to drop them but I have not yet succeeded 😛 

Re-running build ...

zuk3975 committed:
Minor wording fixes

zuk3975 commented:
@matks last requested changes were applied. Could you please restart travis once again to merge this?

matks commented:
> Yes, these tests are flawed 😞
> I'm working hard to convince my teammates to drop them but I have not yet succeeded 😛

Note: I won

https://github.com/PrestaShop/PrestaShop/pull/15161

matks commented:
Now PrettyCI job status has been lost 😡

matks commented:
@zuk3975 All CI checks are now 💚but we need a rebase to fix git conflicts ... 😞

zuk3975 committed:
Merge branch 'develop' into behat/tax

zuk3975 commented:
> @zuk3975 All CI checks are now but we need a rebase to fix git conflicts ...

Resolved conflicts, prettyCI seems to be broken again :smile:

matks commented:
Thank you @zuk3975
